pkg/playground/react-demo-file-upload.tsx

```diff
@@ -0,0 +1,140 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React, { useRef, useState } from "react";
+import { Container, createRoot } from 'react-dom/client';
+
+import { Alert, AlertActionCloseButton } from "@patternfly/react-core/dist/esm/components/Alert/index.js";
+import { Button } from "@patternfly/react-core/dist/esm/components/Button/index.js";
+import { Flex } from "@patternfly/react-core/dist/esm/layouts/Flex/index.js";
+import { Progress } from "@patternfly/react-core/dist/esm/components/Progress/index.js";
+import { TimesIcon, UploadIcon } from "@patternfly/react-icons";
+
+import { FileAutoComplete } from "cockpit-components-file-autocomplete.jsx";
+import { upload } from "cockpit-upload-helper";
+
+const _ = cockpit.gettext;
+
+export const UploadButton = () => {
+    const ref = useRef<HTMLInputElement>(null);
+    const [files, setFiles] = useState<{[name: string]: {file: File, progress: number, cancel:() => void}}>({});
+    const [alert, setAlert] = useState<{variant: "warning" | "danger", title: string, message: string} | null>(null);
+    const [dest, setDest] = useState("/home/admin/");
+    let next_progress = 0;
+
+    const handleClick = () => {
+        if (ref.current) {
+            ref.current.click();
+        }
+    };
+
+    const onUpload = async (event: React.ChangeEvent<HTMLInputElement>) => {
+        cockpit.assert(event.target.files, "not an <input type='file'>?");
+        setAlert(null);
+        await Promise.allSettled(Array.from(event.target.files).map(async (file: File) => {
+            const destination = `${dest}${file.name}`;
+            const abort = new AbortController();
+
+            setFiles(oldFiles => {
+                return {
+                    [file.name]: { file, progress: 0, cancel: () => abort.abort() },
+                    ...oldFiles,
+                };
+            });
+
+            try {
+                await upload(destination, file.stream(), (progress) => {
+                    if (performance.now() < next_progress)
+                        return;
+                    next_progress = performance.now() + 200; // only rerender every 200ms
```

Comment:
Not sure how expensive that is, but perhaps save the now() result in a const? Then you don't need to call it a second time here. (Again, this is model code and likely to be copied around)

Reply:
Ya, I thought about that, but these performance counter type things are usually pretty fast.

I like it from the standpoint of consistency, though.

Reply:
I hope this does not get copy pasted, ideally the uploader in files would be re-usable and put back in cockpit (but let's see first). This is a weird way of debouncing or rendering every 200 ms, maybe a better API is https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame

Seeing that the other code has `XXX` I wonder why this doesn't.

Reply:
The framerate is too frequent here... I don't think causing a react invalidation/re-render is what this function was meant to be used for...